Implement createStatefulListenersMixin

novemberborn opened the pull request:
Fixes #14.

kitsonk commented:
I see the test cases for `setState` but I am a little concerned, maybe just being overly cautious...

Currently it goes like this:

```typescript
{ listeners: { foo: 'action1' } }
/* to */
{ listeners: { foo: 'action1', bar: [ 'action2' ] } }
/* to */
{ listeners: undefined }
```

Where as, I wouldn't mind seeing a modification to that:

```typescript
{ listeners: { foo: 'action1' } }
/* to */
{ listeners: { foo: 'action1', bar: [ 'action2' ] } }
/* to */
{ listeners: { foo: undefined, bar: [ 'action2', 'action3' ] } }
/* to */
{ listeners: undefined }
```

I think it should be fine, because instead of "diffing" the listeners, you simply remove them and then attach whatever is there, but it would be good to have some sort of "deletion" mutation before just the whole deletion.

novemberborn commented:
@kitsonk should we guard against a race condition where `listeners` is changed twice, but the action promises for the second change are resolved *before* those of the first change? The first change will end up overriding the second. Same goes for `StatefulChildrenMixin`.

kitsonk commented:
@novemberborn yes... we should come up with a test for it... like a delay in the registry and then implement some sort of guard on the instance that says it is already waiting to resolve.  Is there any problems simply having the manage be a `then()` chained to the last instance promise for managing the state changes?

novemberborn commented:
@kitsonk pushed several fixup commits (will squash before merging).

* Added the refined test for `setState()`
* Now emitting an `error` event
* Refactoring and test to guard against race conditions (last state wins)

kitsonk commented on src/mixins/createStatefulListenersMixin.ts:
```diff
@@ -0,0 +1,158 @@
+		});
+	}, undefined);
+
+	const replaceListeners = () => {
```
Any particular reason to declare this as a `const` instead of just a function?

novemberborn replied:
Nope. Could change it to a function if you prefer?

kitsonk commented on src/mixins/createStatefulListenersMixin.ts:
```diff
@@ -0,0 +1,158 @@
+	const generation = state.generation = (state.generation || 0) + 1;
+
+	// Unsetting listeners causes previous listeners to be removed.
+	const listeners = evt.state.listeners || {};
```
Minor point, but should we be explicit about the type here?  We are depending upon TypeScript to contextually infer the type from the first value of the statement and the second value's type is overly broad.  TypeScript is "magically" saving us here.

kitsonk commented on src/mixins/createStatefulListenersMixin.ts:
```diff
@@ -0,0 +1,158 @@
+	const widget: StatefulListeners<StatefulListenersState> = <any> evt.target;
+
+	// Assume this function cannot be called without the widget being in the management map.
+	const state = managementMap.get(widget);
```
Slightly confusing variable name, in parts I thought we were referring to a reference of the `widget.state` when in fact, we are simply referring to some internal state.

novemberborn replied:
Will rename to `internalState`.

kitsonk commented on src/mixins/createStatefulListenersMixin.ts:
```diff
@@ -0,0 +1,158 @@
+	(options?: StatefulListenersOptions<StatefulListenersState>): Stateful<StatefulListenersState>;
+}
+
+function resolveListeners(registry: ActionRegistry, cache: Map<string | symbol, Actionable<TargettedEventObject>>, ref: ListenerOrArray): { value?: any; promise?: Promise<any>; } {
```
I guess I am "comfortable" with this, but the TypeScript convention is to use a tuple for multiple returns:

```typescript
function resolveListeners(registry: ActionRegistry, cache: Map<string | symbol, Actionable<TargettedEventObject>>, ref: ListenerOrArray): [ any, Promise<any> ]
```

I guess do we want to stick with TypeScript convention or try to do something different.  I would say unless there is a compelling reasons to break with convention...

kitsonk commented:
> Now emitting an `error` event

Thanks, I remembered that when I was away from the computer this morning and then forgot I remembered it.

novemberborn commented:
@kitsonk:

* Added explicit `ListenersMap` typing
* Renamed `state` variable to `internalState`
* Refactored `resolveListeners()` to return a tuple, and actually type the return values rather than using `any`

kitsonk commented:
👍 

(Also, would be fine if you wanted to comment out Safari 9 on SauceLabs for a while)
